### ISSUE TITLE
ibm.power_aix.devices requires knowledge of current device state to know outcome of task

### ISSUE THREAD

**Issue by stephenulmer** [IBM/ansible-power-aix#98]
**Describe the bug**
ibm.power_aix.devices is capable of updating the attributes of a device, but it is necessary to know the *current* state of the device to know what the result of the task will be.

**To Reproduce**
The following task will silently *not* set the IP address on the interface if it is in the "defined" state, but will instead make the interface available. A following run would then yield different behavior and a different finished state.

```
    - ibm.power_aix.devices:
        device: "en4"
        state: available
        attributes:
          netaddr: "10.100.10.50"
          netmask: "255.255.255.0"
```

**Expected behavior**
The module should make the user's assertion about the state and attributes of a device true.

**Environment:**
 - OS: AIX 7.2TL5
 - Python Version: 2.7
 - AIX Collection Version: HEAD -> dev-collection

**Comment by pvtorres** [comment]
Thanks for helping us to improve the modules Stephen. 
Stephen, did you try to use the mktcpip module?
AIX uses the mktcpip command, which has a corresponding mktcpip module in the collections.
The module will bring up the interface and also it will modify the database attributes (ODM).
In command lines, the actions can be taken by using the command o modify the attributes but then, the ifconfig command will have to be run to bring up and set the values of the running interface. mktcpip takes care of this.

  - name: Configure Network Interface
    mktcpip:
        interface: en1
        address: 10.100.10.50
        netmask: 255.255.255.0
        hostname: <hostname required>

**Comment by stephenulmer** [comment]
This is the second interface, and it's also a VLAN interface. I'm pretty sure I failed to get that combination to do what I meant with mktcpip, but I could have just been doing it wrong.

Looking at what I decided to do for the play, I think that there was no way to actually do a mkdev, which I needed to make the VLAN device (entX), and I might have just gotten on the device track from trying to figure that out -- I'm doing that with the command module right now. I'm also using */usr/lib/methods/defif* to make the actual interfaces({en,et}X), which feels dirty.

**Comment by pvtorres** [comment]
The mktcpip will take care of doing all the rmdv/mkdev and chdev commands as the interfaces are not pretty straight forward on how the attributes are changed. Let me know if this works and if it does, I can close the issue.

**Comment by stephenulmer** [comment]
I will make another run at this and let you know if I can make that work. However, I'm **not** asking for *devices* to sanitize or check the validity of the attributes, that is definitely the user's responsibility.

I am asking for *devices* to be idempotent. If we run it twice in a row it may change/do different things for each run based on the state of the device we started with.

The problem is that the module literally ignores the attributes sometimes. Those attributes can be set (at least in the ODM) no matter the state of the device. If the module allows changing the state of the device, then the state of the device should be what the user specifies after the module is finished. The module should not sometimes do some of the things depending on the device's state when the module is run.  Ansible doesn't have a "partially successful" or "mostly changed" result state.

Maybe this should be two modules, one that changes the device state and one that sets attributes?

**Comment by stephenulmer** [comment]
My resources got moved around, but I should be able to try this again very soon.

**Comment by pvtorres** [comment]
Stephen, I see what you ware seen in the module. The module assumes the user knows the state of the device, and if the device does not match the state, it will ignore the changes. 
Also it seems that it assumes the device attributes can be changed without disrupting the device. The module sends 
the change command attributes with flags that indicate that, no disruption. But there are some devices that requires a re-configuration and this devices will reject the change of the attributes. 
So there are two issues there. 
I will work on a fix for it where the module will just execute the change of the attributes regarding of the state of the device unless the user specifies with a existing attribute, that the change should be after reboot.  Let me know if you agree with this.

**Comment by stephenulmer** [comment]
I am so very sorry it took me this long to respond!

Yes, I agree with what you’re saying about the current state, and I think I agree with the solution. In my opinion the user is responsible for understanding what can and can’t be changed about the device without disruption, so the user should specify the ODM-only-equivalent flag if necessary.

In all cases, the module should either do what the user asks it to, or fail because it can’t. In no cases should the module provide different real-world results on subsequent runs with the same input.

-- 
Stephen

**Comment by pvtorres** [comment]
@stephenulmer Take a look of the fixes. I was able to change the attributes of the en0 interface and see them reflected. I did multiple fixes for idempotency issues, removal recursive issues, etc. The module does not depend on the state of the device to execute the commands (like changing the attributes). Also, there are some devices that does not support the dynamic attribute changes (changes while the adapter is left in Available state). Those adapters need a reconfiguration when the attributes are changed. If such devices need the attributes changes, you can use the chtype = reset option and it will take care of it. (example, cryptographic adapters or the pkcs11 pseudo device).

**Comment by stephenulmer** [comment]
Okay, I finally got to test this today!

It now does what I was expecting for all of the situations I have coded, and it's also idempotent for all of my test cases.

Thank you very much!

**Comment by stephenulmer** [comment]
I did some additional testing today and discovered a small problem:

If the user does *not* specify a device at all (don't ask how I managed to do this), then the module fails because eventually it tries to run:

```lsattr -El  all```

one the target. Now this only happens if you screw up the task in the YAML in the first place, but it seems worth mentioning.

As far as I am concerned, the change is appropriate and meets my needs.
